packages/suite/src/actions/wallet/send/sendFormSolanaActions.ts

```diff
@@ -1,13 +1,145 @@
 import { Transaction, SystemProgram, PublicKey } from '@solana/web3.js';
 import {
     FormState,
     PrecomposedTransactionFinal,
+    ComposeActionContext,
+    ExternalOutput,
+    PrecomposedTransaction,
+    PrecomposedLevels,
 } from '@suite-common/wallet-types';
 import { notificationsActions } from '@suite-common/toast-notifications';
 import { Dispatch, GetState } from 'src/types/suite';
 import {
+    calculateMax,
+    calculateSolFee,
+    calculateTotal,
+    formatAmount,
+    getExternalComposeOutput,
     getLamportsFromSol,
 } from '@suite-common/wallet-utils';
+import BigNumber from 'bignumber.js';
+
+// TODO(vl): Implement send-max
+// TODO(vl), phase 2: Implement custom Prioritization Feetat
+
+// TODO(vl): Handle token transfers -- see ETH implementation
+const calculate = (
+    availableBalance: string,
+    output: ExternalOutput,
+    feeLevel: FeeLevel,
+    token?: TokenInfo,
+): PrecomposedTransaction => {
+    // TODO(vl): handle prioritisation fee
+    // 1, 0, 0 = discard prioritisation fee for now
+    // Default args include the prioritisation fee
+    const feeInSatoshi = calculateSolFee(1, 0, 0);
+    let amount: string;
+    let max: string | undefined;
+    if (output.type === 'send-max' || output.type === 'send-max-noaddress') {
+        max = calculateMax(availableBalance, feeInSatoshi);
+        amount = max;
+    } else {
+        amount = output.amount;
+    }
+
+    // total SOL spent (amount + fee), in case of SPL token only the fee
+    const totalSpent = new BigNumber(calculateTotal(token ? '0' : amount, feeInSatoshi));
+
+    if (totalSpent.isGreaterThan(availableBalance)) {
+        const error = token ? 'AMOUNT_NOT_ENOUGH_CURRENCY_FEE' : 'AMOUNT_IS_NOT_ENOUGH';
+        // errorMessage declared later
+        return { type: 'error', error, errorMessage: { id: error } } as const;
+    }
+
+    const payloadData = {
+        type: 'nonfinal',
+        totalSpent: token ? amount : totalSpent.toString(),
+        max,
+        fee: feeInSatoshi,
+        feePerByte: feeLevel.feePerUnit,
+        feeLimit: feeLevel.feeLimit,
+        token,
+        bytes: 0,
+    } as const;
+
+    if (output.type === 'send-max' || output.type === 'external') {
+        return {
+            ...payloadData,
+            type: 'final',
+            // compatibility with BTC PrecomposedTransaction from @trezor/connect
+            transaction: {
+                inputs: [],
+                outputsPermutation: [0],
+                outputs: [
+                    {
+                        address: output.address,
+                        amount,
+                        script_type: 'PAYTOADDRESS',
+                    },
+                ],
+            },
+        };
+    }
+
+    return payloadData;
+};
+
+export const composeTransaction =
+    (formValues: FormState, formState: ComposeActionContext) => () => {
+        const { account, network, feeInfo } = formState;
```

Comment:
I know you might address this in a separate PR, but what is `feeInfo` for solana? What it includes now and what will it include later? If currently do not support custom fees, I would remove all logic that handles that, otherwise its a code that needs to be reviewed but cannot be tested or reasoned about since its irrelevant.

Reply:
Right now it includes only the normal fee. I felt it was valuable to leave this custom fee scaffolding as it is not going to change (same as other chain's impl) and will make custom fees work out of the box once the UI can handle them. Similarly, adding different default fee levels once we decide on them should be handled automatically here.

Reply:
Sure I get it, lets say we perfectly know now what's the implementation is going to be. The thing is, if we introduce such a logic now, and then dismiss it saying that "its not used now but will be", it could lead the reviewer to basically skip this part. In the next PR, in which this logic will be used, this part wont be among changes, so it will not be considered. For this reason and for couple of other which are too long to explain here, I would still be for removing anything that is not used in the current implementation, and rather reintroduce it once its used. 

I am not sure to what extend this is the case right now, maybe its nothing, or just minor stuff, still I would keep just what its currently used even tho we now know that we will reintroduce it again later.

Reply:
Understandable, updated.